### ISSUE TITLE
[Bug]: Unreliable page closing in the presence of a prompt

### ISSUE THREAD

**Issue by ThibTrip** [microsoft/playwright#10476]
### Playwright version

1.17.0

### Operating system

Linux

### What browsers are you seeing the problem on?

Chromium

### Other information

OS: Pop!_OS 21.04
Python: 3.9.7

### What happened? / Describe the bug

Hi,

I am trying to use the Chromium browser with **each tab controlled asynchronously** (I simplified the code as much as possible, my project is a fully async crawler with URLs feed from a PostgreSQL queue using the new sqlalchemy async engines).

I noticed that if I land on **a page with a prompt** (user/password) **sometimes it will not get closed**.

### Code snippet to reproduce your bug

```shell
# # Description
#
# This shows a potential problem with the playwright library with tabs remaining forever despite closing them.
# This seems to happen somewhat randomly and only in the presence of prompts.
import asyncio
import sys
import threading
import traceback
from asyncio import TimeoutError as AsyncTimeoutError
from playwright.async_api import async_playwright

# # Config

MAXIMUM_CONCURRENCY = 7 # urls in parallel
TIMEOUT_URL = 10 # seconds
NB_REPEATS = 5 # we will call the same URLs multiple times as errors are a bit random
HEADLESS = False

# # Data

urls = ['https://en-master.skoda-auto.com/models/scala',
        'https://cs-cz-260-22222.cpv3prod.skoda-auto.com/prodej-novych-vozu/modely',
        'https://en-master-v2.skoda-auto.com/_doc/cfb6fbbd-3a21-4cd6-a52d-e81cbb7f5fa9',
        'https://en-master-v2.skoda-auto.com/_doc/eea2f7ce-ee89-42cf-959c-21fe73a133ac',
        'https://en-master.skoda-auto.com/company/myskoda-iv-app',
        'https://en-master.skoda-auto.com/models/fabia/fabia-combi-scoutline/fabia-combi-scoutline']*NB_REPEATS

# # Helpers and workers

semaphor = asyncio.Semaphore(MAXIMUM_CONCURRENCY) # limit concurrency

async def get_browser_context():
    p = await async_playwright().start()
    playwright_browser = await p.chromium.launch(headless=HEADLESS)
    return await playwright_browser.new_context()

async def feeder(queue):
    """
    Fetches URLs from some source and puts them into a queue for our other workers
    """
    for url in urls:
        queue.put_nowait(url) # put an item into the queue without blocking
        await asyncio.sleep(0.05) # don't spam the queue

async def downloader(queue, context):
    """
    Fetches the content of a webpage
    """
    while True:
        # no more jobs? try again later
        if queue.empty():
            await asyncio.sleep(0.05)
            continue

        # set page to a default in case we cannot create one for some reason
        page = None 
        try:
            await semaphor.acquire()
            url = await queue.get()
            loop = asyncio.get_event_loop()
            page = await context.new_page()
            # use timeout in playwright and then another longer timeout for the coroutine
            # in case something gets stuck so we are sure nothing keeps running forever
            # and pages are always closed
            coro = page.goto(url, timeout=TIMEOUT_URL*1000)
            result = await asyncio.wait_for(coro, timeout=20)
            content = await response.body()
        except Exception:
            traceback.print_exc()
        finally:
            if page is not None:
                await page.close()
            semaphor.release()
            queue.task_done()

# # Main

async def main():
    loop = asyncio.get_event_loop()
    queue = asyncio.PriorityQueue()
    context = await get_browser_context()
    await asyncio.gather(feeder(queue),
                         asyncio.gather(*[downloader(queue=queue, context=context) for i in range(MAXIMUM_CONCURRENCY)]),
                         return_exceptions=False)

# # Run

if __name__ == '__main__':
    asyncio.run(main(), debug=True)
```

**Comment by ThibTrip** [comment]
> If the page requires authentication and there are no credentials, page.goto will result in 401 error code without showing any prompts. Can you reduce you test case to a single failing url?

That's the thing sometimes it closes, sometimes it does not which is why I even included repetition 😐

**Comment by yury-s** [comment]
If I insert a delay before next iteration in headful mode I do see the http credentials dialog pop up. Otherwise without the delay the page close before the dialogs appear, probably some race condition related to that.

**Comment by yury-s** [comment]
For the record, it is only reproducible in headful mode.

**Comment by mxschmitt** [comment]
Seems like `Target.detachedFromTarget` CDP does not get emitted in all the cases and it hangs when a dialog is open.

As a workaround you can launch your browsers headless which should work.

**Comment by ThibTrip** [comment]
Good to know that this does not happen headless :+1: , thanks!
I am only running headful for testing actually :relaxed: (it helps to see how the browser is behaving) so as far as I am concerned this is a non issue (does not mean this issue should be closed, after all there seems to a bug, but that is up to you :see_no_evil:).

**Comment by mxschmitt** [comment]
We definitely won't close it. Mostly we work on the demand of an issue, which means we count upvotes. Also in this particular case, it's an upstream Chromium bug, which usually takes more time for us to invest compared to Playwright bugs.

**Comment by mxschmitt** [comment]
This should be fixed in master by https://github.com/microsoft/playwright/pull/11614.

**Comment by aslushnikov** [comment]
I think #11614 only fixes part of the issue caused by JS Dialogs; others (like HTTP auth) might still be a problem. So let's keep it open for now!

**Comment by pavelfeldman** [comment]
#### Why was this issue closed?

We are prioritizing the bugs based on the upvotes, recency and our ability to act. It looks like this issue only has a handful of upvotes, has not been touched recently and/or we lack sufficient feedback to act on it. We are closing issues like this one to keep our bug database maintainable. Please feel free to open a new issue and link this one to it if you think this is a mistake.
